docs/Validation.md

```diff
@@ -320,42 +320,46 @@ const CustomerCreate = () => (
 
 ## Server-Side Validation
 
-You can use the errors returned by the dataProvider mutation as a source for the validation. In order to display the validation errors, a custom `save` function needs to be used:
+Server-side validation is supported out of the box. It requires that the dataProvider throws an error with the following shape:
 
-{% raw %}
-```jsx
-import * as React from 'react';
-import { useCallback } from 'react';
-import { Create, SimpleForm, TextInput, useCreate } from 'react-admin';
-
-export const UserCreate = () => {
-    const [create] = useCreate();
-    const save = useCallback(
-        async values => {
-            try {
-                await create('users', { data: values }, { returnPromise: true });
-            } catch (error) {
-                if (error.body.errors) {
-                    // The shape of the returned validation errors must match the shape of the form
-                    return error.body.errors;
-                }
-            }
-        },
-        [create]
-    );
-
-    return (
-        <Create>
-            <SimpleForm onSubmit={save}>
-                <TextInput label="First Name" source="firstName" />
-                <TextInput label="Age" source="age" />
-            </SimpleForm>
-        </Create>
-    );
-};
 ```
-{% endraw %}
+{
+    body: {
```

Comment:
why the "body" key? Why can't the dataProvider only reject with `({ errors: { source: 'error message' } }`?

Also, your snippet is both a type description and an example, and so it's not enough informative. I'd use a more complete example:

```
{ 
  errors: {
    title: 'An article with this title already exists. The title must be unique.',
    date: 'The date is required',
    tags: { message: "The tag 'agrriculture' doesn't exist" },
  }
}
```

Reply:
Because that's the current shape of our `HttpError`. Our data providers put the response json inside the `body` property of the error they throw